pkg/resources/validation/cluster/cluster.go

```diff
@@ -93,16 +154,136 @@ func (v *Validator) Admit(request *admission.Request) (*admissionv1.AdmissionRes
+	oldTemplateName := oldCluster.Spec.DefaultPodSecurityAdmissionConfigurationTemplateName
+	parsedVersion, err := getClusterVersion(newCluster.Spec.RancherKubernetesEngineConfig.Version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cluster version: %w", err)
+	}
+	if parsedRangeLessThan125(parsedVersion) && newTemplateName != "" {
+		msg := "PodSecurityAdmissionConfigurationTemplate(PSACT) is only supported in Kubernetes version 1.25 and above"
+		return psa.AdmissionResponseBadRequest(msg), nil
+	}
+	if newTemplateName != "" {
+		response, err := v.checkPSAConfigOnCluster(newCluster)
+		if err != nil {
+			return nil, fmt.Errorf("failed to check the PodSecurity Config in the cluster %s: %w", newCluster.Name, err)
+		}
+		if !response.Allowed {
+			return response, nil
+		}
+	} else {
+		switch request.Operation {
+		case admissionv1.Create:
+			return psa.AdmissionResponseAllowed(), nil
+		case admissionv1.Update:
+			// In the case of unsetting DefaultPodSecurityAdmissionConfigurationTemplateName,
+			// validate that the configuration for PodSecurityAdmission under the kube-api.admission_configuration section
+			// is different between the new and old clusters.
+			// It is possible that user drops DefaultPodSecurityAdmissionConfigurationTemplateName and set the config
+			// under kube-api.admission_configuration at the same time.
+			if oldTemplateName != "" {
+				newConfig, found := psa.GetPluginConfigFromCluster(newCluster)
+				if !found {
+					// not found means the kube-api.admission_configuration section is also removed, which is good
+					return psa.AdmissionResponseAllowed(), nil
+				}
+				oldConfig, _ := psa.GetPluginConfigFromCluster(oldCluster)
+				if reflect.DeepEqual(newConfig, oldConfig) {
+					msg := "The Plugin Config for PodSecurity under kube-api.admission_configuration is the same as the previously-set PodSecurityAdmissionConfigurationTemplate." +
+						" Please either change the Plugin Config or set the DefaultPodSecurityAdmissionConfigurationTemplateName."
+					return psa.AdmissionResponseBadRequest(msg), nil
+				}
+			}
+		}
+	}
+	return psa.AdmissionResponseAllowed(), nil
 }
 
-func toExtra(extra map[string]authenticationv1.ExtraValue) map[string]v1.ExtraValue {
-	result := map[string]v1.ExtraValue{}
-	for k, v := range extra {
-		result[k] = v1.ExtraValue(v)
+// checkPSAConfigOnCluster validates the cluster spec when DefaultPodSecurityAdmissionConfigurationTemplateName is set
+func (v *Validator) checkPSAConfigOnCluster(cluster *apisv3.Cluster) (*admissionv1.AdmissionResponse, error) {
+	// validate that extra_args.admission-control-config-file is not set at the same time
+	_, found := cluster.Spec.RancherKubernetesEngineConfig.Services.KubeAPI.ExtraArgs["admission-control-config-file"]
+	if found {
+		msg := "could not use external admission control configuration file when using PodSecurityAdmissionConfigurationTemplate"
```

Comment:
Nit, but why do we set `msg` var instead of just returning `psa.AdmissionResponseBadRequest("...message...")`?

Reply:
nothing special but to avoid a long return line, but I feel it does not help much.